### DISCUSSION TITLE
Camel Karavan vs. Camel-K: Did I get it wrong?

### DISCUSSION THREAD

**Discussion by TheHandOfNOD** [apache/camel-karavan#1157] [General]
Hi guys,
I wanted to instrumentalize Karavan as an Tool to develop integrations which will then be deployed via Camel-K-Operator into my Kubernetes. This initiative was driven by the several documentation resources which I might have understood wrongly because what I figured out after installing and using Karavan is that Karavan actually only uses JBang and builds by itself a Dockerimage which is then used to create a container.

I would have thought that after "designing" my Route and combining all my Kamelets, somehow the "Kamel CLI" will be used to actually create a Camel-K integration.

So my questions are: 
- Was it never intended that Karavan works seamlessly with the Camel-K-Operator?
- Why is this the case?
- Do have to make a choice between Karavan and Camel-K? 

Unfortunately for my Setup it would not possible to use the VS-Code-Karavan-Plugin and use the Kamel-CLI manually to achieve this, but rather I have to use only Karavan.
Would love if someone could bring light into this topic for me.
Thx in advance.

**Comment by mgubaidullin** [comment]
1. Karavan does not work with Camel-K-Operator

2. There are 2 options to use Karavan:
- vs-code extension
- web application in k8s 

Both options allow you to run integration application, build container and deploy to k8s.
Which option do you prefer?

**Reply by jwgmeligmeyling** [reply]
I get the "it's a different project" standpoint but from a user standpoint the differences can be a bit surprising. Both are Kubernetes based projects under the Camel K umbrella.

When running as Karavan as web application in K8s, couldn't Karavan benefit from Camel-K somehow? Having both Camel-K and Karavan deployed in Kubernetes, I was a bit surprised they didn't share the catalog, build mechanism (Tekton/Jbang) and deployment mechanism (deployments vs integrations). I think Camel K could benefit just as well from a UI ( I am thinking something like Red Hat Integration - Camel K).

Maybe I am missing the point of Kubernetes backed Karavan still, but why the wouldn't Karavan be able to produce Camel K resources as output?

**Reply by TheHandOfNOD** [reply]
@jwgmeligmeyling : I thought the same. It's quite irritating that these 2 projects do not complement each other in any way. To enforce this confusion you can even select Kamelets in Karavan. 

I think Karavan might can do more than Camel-K (even though I am not sure) but when I want to enable a Citizen Integrator I cannot choose any of these tools now.
Camel-K is to developer heavy to be able to build something for the citizen integrator
Karavan, which I could leverage as No-Code/Low-Code Platform for my Team who are not full devs, is not able to instrumentalize Camel-K.

**Reply by mgubaidullin** [reply]
0. **Karavan** and **Camel K** are both under **Camel** umbrella. 
1. Karavan works on **Docker** or k8s. Other container management systems considered
2. Karavan stores code in git and deploys containers, while Camel-K requires code to be applied as resources to k8s
3. Karavan is written in Java while Camel-K Operator in Go
4. Karavan provides full control on build and deployment process (build.sh)
5. Karavan aimed to provide full developer experience for low-code integration development (design, build and deploy)
6. Karavan uses Camel-main and does not require Quarkus or SpringBoot

In Apache Camel community we have build runtimes: **Camel-main**, **Quarkus**, **Spting-Boot**, **Karaf**, multiple DSLs: **Java**, **XML**, **YAML**. Same way **Karavan** and **Camel-K** (while also having some overlap functionality) are enriching Apache Camel ecosystem by providing alternative ways of building and managing integration services with Apache Camel.

**Reply by TheHandOfNOD** [reply]
Hi,

I think you might have got my question wrong and sorry for the late reply but it wasn't easy to find the resource which lead me into thinking that Karavan uses Camel-K internally to create the Integration Custom Resources:

https://camel.apache.org/blog/2021/10/camel-karavan-preview-release/

 I quote:

`After you finish to configure the integration, the output you get from the UI is an Integration custom resource, that you can direcly apply to any Kubernetes cluster (having Camel K installed) and it will automatically run.`

This seems not to be true (anymore?) as I only get a route.yaml file and the build button uses the build-docker.sh file in which a container image is created and not an integration custom resource.

Surely the following blog posts and videos fake the cooperation of Karavan and Camel-K because only the VS-Code-Plugin of Karavan is used to create the yaml but the CommandLine is used with the Kamel K operator.

My question was just: Why doesn't Karavan offer the option to use Camel K Operator in the first place?

**Reply by davsclaus** [reply]
This is an old blog post in the initial days of the Karavan experimentation.
If you want to read and understand about Karavan then check its website and latest documentation, and latest blog posts.

**Reply by davsclaus** [reply]
You can customize karavan anyhow you like, as you have discover those build files, such as [`kuber-docker.sh`](https://github.com/apache/camel-karavan/blob/main/karavan-app/src/main/resources/snippets/kubernetes-build.sh) then you can change it anyhow you like - if you have Camel K installed, then just use its Kamel CLI to install.

**Comment by davsclaus** [comment]
Karavan is using vanilla standard Apache Camel (camel-main) and have no intention to do Camel K.
